pkg/guacanalytics/patchPlanning.go

```diff
@@ -0,0 +1,187 @@
+//
+// Copyright 2023 The GUAC Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package guacanalytics
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Khan/genqlient/graphql"
+	model "github.com/guacsec/guac/pkg/assembler/clients/generated"
+	"github.com/guacsec/guac/pkg/misc/depversion"
+)
+
+type DfsNode struct {
+	Expanded     bool // true once all node neighbors are added to queue
+	Parent       string
+	Depth        int
+	NodeType     string   // packageName, packageVersion
+	nodeVersions []string // for a packageName, what was the packageVersion associated with this version
+}
+
+type queueValues struct {
+	nodeMap map[string]DfsNode
+	now     string
+	nowNode DfsNode
+	queue   []string
+}
+
+// TODO: make more robust using predicates
+func SearchDependenciesFromStartNode(ctx context.Context, gqlclient graphql.Client, startID string, stopID *string, maxDepth int) (map[string]DfsNode, error) {
+	startNode, err := model.Node(ctx, gqlclient, startID)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed getting intial node with given ID:%w", err)
+	}
+
+	nodePkg, ok := startNode.Node.(*model.NodeNodePackage)
+
+	if !ok {
+		return nil, fmt.Errorf("Not a package")
+	}
+
+	q := queueValues{
+		queue:   make([]string, 0), // the queue of nodes in bfs
+		nodeMap: map[string]DfsNode{},
+	}
+
+	// TODO: add functionality to start with other nodes?
+	if len(nodePkg.AllPkgTree.Namespaces) < 1 {
+		return nil, fmt.Errorf("Start by inputting a packageName or packageVerion node")
+	}
+
+	if len(nodePkg.AllPkgTree.Namespaces[0].Names) < 1 {
+		return nil, fmt.Errorf("Start by inputting a packageName or packageVerion node")
+	}
+
+	if len(nodePkg.AllPkgTree.Namespaces[0].Names[0].Versions) < 1 {
+		var versionsList []string
+		for _, versionEntry := range nodePkg.AllPkgTree.Namespaces[0].Names[0].Versions {
+			versionsList = append(versionsList, versionEntry.Version)
+		}
+		q.nodeMap[startID] = DfsNode{
+			NodeType:     "packageName",
+			nodeVersions: versionsList,
```

Comment:
nit: add a TODO here to note that it needs to handle the case where there's circular depedencies that introduce more versions to the version list on a node that requires revisiting